### ISSUE TITLE
enhancement: add azurerm provider linting

### ISSUE THREAD

**Issue by wardcomm** [wata727/tflint#317]
enhancement: add azurerm provider linting

**Comment by wata727** [comment]
I think this is a valid request, but I now focused on aws provider linting. Unfortunately, there are no specific plans.

Duplicate of #175

**Comment by jpreese** [comment]
Out of curiosity, is adding additional providers something you do want to support? Looking through some of the implementations / help menus, it does seem focused on AWS. Which is understandable. 

I'm curious if your plan is to support all providers, or if you think it'd be better to split it out into another project. It would be a much larger effort, but I almost wonder if it'd be better to have tflint be a baseline tf linter and each provider is its own plugin for tflint.

We are just now doing a heavy Terraform push (using Azure) and are very interested in linting capabilities. We'd be interested in adding rules we stumble upon, but I don't want to try and shove Azure down tflint's throat if its just not a good approach.

**Comment by wata727** [comment]
@jpreese 

>  I almost wonder if it'd be better to have tflint be a baseline tf linter and each provider is its own plugin for tflint.

Yeah, I totally agree. The idea of the plug-in system is now developed in #255.
However, TFLint's internal API is not yet stable, and breaking changes may be introduced to the plug-in interface frequently. Therefore, I'm a little concerned that this system will be widely used in the early stage.

> We'd be interested in adding rules we stumble upon

Thank you for your interest. It's very worthwhile to add rules that you think is informative and I welcome it.
In the future, I'm thinking about cutting them into plugins, but at this point, it would be a good idea to just add them to this repository.

The only concern is how to keep those rules up to date. As I'm not familiar with Azure, it may not always be possible to reflect the latest updates.
In order to solve this problem, I think it is necessary to build a system that can automatically detect that constraints have been updated, but this is another issue.

**Comment by ashemedai** [comment]
A lot of that documentation can either be found via https://docs.microsoft.com/en-us/azure/templates/, https://github.com/Azure/azure-resource-manager-schemas, or direct queries via the az cli. So a large portion could be automated.

**Comment by wata727** [comment]
Excellent! I'm working on implementing the plug-in system now, but I'm happy if you help to implement Azure rules as plug-ins after that.
